### PR TITLE
lp1783392: Configurable debouncing timeout for WSearchLineEdit

### PR DESCRIPTION
**Pull request by uklotzde**
Add a configurable debouncing timeout for the search edit field. Depending on the size of their library users are now able to adjust this to mitigate the issues caused by locking the UI thread while searching.

Another user now has exactly reported the same issues that I am experiencing, i.e. Mixxx is almost unusable with bigger libraries and the default settings: https://bugs.launchpad.net/mixxx/+bug/1783392

### CODE REVIEW

File: src/preferences/dialog/dlgpreflibrary.cpp [hunk 1]
```diff
@@ -11,8 +11,11 @@
 #include "preferences/dialog/dlgpreflibrary.h"
 #include "library/dlgtrackmetadataexport.h"
 #include "sources/soundsourceproxy.h"
+#include "widget/wsearchlineedit.h"
 
-#define MIXXX_ADDONS_URL "http://www.mixxx.org/wiki/doku.php/add-ons"
```
Comment (daschuer) [inline src/preferences/dialog/dlgpreflibrary.cpp]: This macro is still in use. Is this an unrelated change, slipped in?

File: src/preferences/dialog/dlgpreflibrarydlg.ui [hunk 2]
```diff
@@ -235,24 +242,34 @@
...
         </property>
        </widget>
       </item>
-      <item row="4" column="0" colspan="2">
-       <widget class="QCheckBox" name="checkBoxEditMetadataSelectedClicked">
+      <item row="5" column="0">
+       <widget class="QLabel" name="incrementalSearchDelayLabel">
         <property name="text">
-         <string>Edit metadata after clicking selected track</string>
+         <string>Incremental Search Delay:</string>
```
Comment (daschuer) [inline src/preferences/dialog/dlgpreflibrarydlg.ui]: This might be misleading for me since "Incremental Search" for me is more an algorithm that works all the time in background guessing the final user result. Google has it in the search bar if the network bandwidth is good enough. 
That would be great for Mixxx, but we are not there.   

"Search after edit" 
"Search As You Type Delay"
Or inverse 
"Search inhibit time"

"Zero disables the feature" ??

File: src/preferences/dialog/dlgpreflibrary.cpp (hunk 1 above)
Comment (uklotzde) [reply src/preferences/dialog/dlgpreflibrary.cpp]: Slipped in. I didn't notice, because I'm already using a version without plugins for development. I've removed this obsolete #define now from master to prevent such mistakes in the future.

File: src/preferences/dialog/dlgpreflibrarydlg.ui (hunk 2 above)
Comment (uklotzde) [reply src/preferences/dialog/dlgpreflibrarydlg.ui]: Ok, I will revert back to the _debouncing_ terminology also for the configuration setting for consistency.

For the text on the label we can use one of the proposed variants. Feedback by native speakers and especially non-developers welcome ;)

**Comment by uklotzde** [comment]
The configurable timeout has proven to be very useful to me, because the optimal setting depends on both the size of your library as well as your personal typing speed and habits.

We have a similar fixed timeout for navigation within the side bar that is currently set to kPressedUntilClickedTimeoutMillis = 300. But this is not so sensitive, because when repeatedly clicking a button or turning an encoder you don't have to account for the delays while looking up different keys or thinking about how to finish your search.

File: src/preferences/dialog/dlgpreflibrarydlg.ui (hunk 2 above)
Comment (uklotzde) [reply src/preferences/dialog/dlgpreflibrarydlg.ui]: I don't want to allow a configurable timeout of 0 to disable this feature entirely. A maximum delay of almost 10 sec should be ok. But I've fixed the code to handle empty or invalid timeouts by not starting the timer.

**Review by daschuer — CHANGES_REQUESTED** [review]
The timeout is bypassed if you delete the last character in the search box. 
This also happens also in 2.1.
Can we have a minimal fix it there as well? 

If I put the settingsPath on a USB HDD, I can reproduce the leaks. But they only happens if many tracks are shown or hidden due the search. 
We may be able to explain it if all tracks are shown again, but why is this happening when tracks are hidden? Do we have a caching issue? 

Disabling some columns doe does not change the issue by the way.

**Comment by daschuer** [comment]
> The setting is now configurable, which might disqualify it from being included in the 2.2 beta.

Unfortunately yes. Maybe the major issue can be fixed via solving the bypass issue. 
A slightly longer delay may also fix some of the issues.

**Comment by uklotzde** [comment]
I cannot confirm that the timeout and timer is bypassed when deleting single characters. Deleting the last character with backspace triggers text changed events and the timer is reset.

**Comment by uklotzde** [comment]
Target branch: One option would be to split the UI changes off from this PR and merge them separately into _master_. The new configuration remains undocumented, but could already be used by editing a single line in the config file manually.

We might need this configuration option to help fixing issues of users with big libraries. I expect at least some complaints once the new version has been released to a broader audience.

**Comment by daschuer** [comment]
Are you able to reproduce the deleting the last character issue?

**Comment by uklotzde** [comment]
Deleting the last characters works as like any other editing operation and restarts the timer as expected.

**Comment by uklotzde** [comment]
With Qt4 I was able to continue typing while the UI was frozen and ended up with the search phrase I typed. With Qt5 the characters typed while the UI is frozen are reordered in unforeseeable ways.

File: src/widget/wsearchlineedit.cpp [hunk 3]
```diff
@@ -1,193 +1,226 @@
...
+void WSearchLineEdit::updatePlaceholder(QString text) {
+    if (text.isEmpty()) {
+        showPlaceholder();
+    } else {
+        hidePlaceholder(text);
+    }
+}
+
+void WSearchLineEdit::updateClearButton(QString text) {
+    m_clearButton->setVisible(!text.isEmpty() && !m_showingPlaceholder);
```
[the remark this answers is not in the document]
Comment (daschuer) [reply src/widget/wsearchlineedit.cpp]: OK, than lets keep it like that.

**Comment by daschuer** [comment]
The issue with the last character is "emit searchCleared();" during typing. 
Can't we remove this signal completely and check for empty in onSearch?

I am also still unsure about the target branch. 
Do we have other users suffering the character mix up issue?

**Comment by uklotzde** [comment]
I misunderstood your description "last character"! Ok, now I got it. Deleting the last character before the edit field becomes empty ;)

**Comment by uklotzde** [comment]
- Another rebase to restore the lost disabledSearch() signal from the previous rebase
- Activate timer upon deletion of last character in field
- Simpler and more robust state handling

It still doesn't feel right, but I'm tired of improving this even more. I will only fix outstanding bugs. No backport to 2.1, this has already cost me much more time then planned.

**Comment by daschuer** [comment]
Thank you. 
What do you mean by:
> It still doesn't feel right?

We also don't have a conclusion if this as suits for 2.2. Do we know if this a Wayland issue? 

I have tried to find something about the key stroke reordering on the web, but I haven't found anything.
Does it mean that this is induvidual an issue of your setup? Or is it sn issue that only happens with a certain Wayland and Qt version? Or is it a simple Mixxx bug?
What do you think?

**Comment by uklotzde** [comment]
"still not right": The underlying state machine with its side effects is still complicated. I'm not able to guarantee that all possible states and transitions are covered and work as expected. Testing with a large library that causes ui freezes and by varying the timeout were hopefully sufficient to reveal and fix most issues.

**Comment by uklotzde** [comment]
I've update the target to master/2.3.0, but I will keep this branch aligned with 2.2 until we know if this is an issue that affects multiple users.

**Review by daschuer — APPROVED** [review]
